src/bitprim/interface/block_chain.cpp

```diff
@@ -80,6 +84,104 @@ void block_chain::for_each_transaction_non_coinbase(size_t from, size_t to, bool
         ++from;
     }
 }
+#ifdef WITH_KEOKEN
+
+void block_chain::get_to_keo_transaction(libbitcoin::hash_digest hash, std::shared_ptr<std::vector<transaction_const_ptr>> keoken_txs) const {
+   fetch_transaction(hash, true, false,
+              [&](const libbitcoin::code &ec,
+                  libbitcoin::transaction_const_ptr tx_ptr, size_t index,
+                  size_t height) {
+                  if (ec == libbitcoin::error::success) {
+                      auto keoken_data = bitprim::keoken::first_keoken_output(*tx_ptr);
+                      if(!keoken_data.empty())
+                          (*keoken_txs).push_back(tx_ptr);
+                  }
+          });
+}
+
+void block_chain::fetch_keoken_history(const short_hash& address_hash, size_t limit,
+    size_t from_height, keoken_history_fetch_handler handler) const
+{
+    auto keoken_txs = std::make_shared<std::vector<transaction_const_ptr>>();
+    if (stopped())
+    {
+        handler(error::service_stopped, keoken_txs);
+        return;
+    }
+
+//TODO RESTRICT MINIMUM HEIGHT (define bitprim::starting_keoken_height)
+/*
+    if(from_height < bitprim::starting_keoken_height)
+        from_height = bitprim::starting_keoken_height;
+*/
+    auto history_compact_list =  database_.history().get(address_hash, limit, from_height);
+
+    for (const auto & history : history_compact_list) {
+        if((*keoken_txs).empty())
+            get_to_keo_transaction(history.point.hash(), keoken_txs);
+        else
+            if(history.point.hash() != keoken_txs->back()->hash())
```

Comment:
what happend if exists tx hash collitions ?

Reply:
If an address has an input and an output in the same transaction, it will be listed twice in the history list. We only want to list each transaction once.